### DISCUSSION TITLE
Lights without using Fog of War?

### DISCUSSION THREAD

**Discussion by PalasX** [Kruptein/PlanarAlly#628] [Q&A]
I think i am misunderstanding how Auras and Lights work when "Fill map with FOW" is disable. I have a Token setup for control by a player, and that token has a Torch Aura, with 20 feet bright light and 20 feet dim light. The "lightbulb" and "eye" are both enabled for this Aura, meaning it is a light source and can be seen by others (aka, its not darkvision).

Everything works as expected while "Fill map with FOW" is enabled. When i disable "Fill map with FOW" or move to a map that is not filled with FOW, i get a "darkness aura" from this token, and i have to turn the light source off to see:
![image](https://user-images.githubusercontent.com/37154966/103949389-2c9d6480-5109-11eb-9a70-ccd884540e7f.png)

When i turn the light source off, i get what you would expect a light source to do in bright daylight, slightly color the area around it:
![image](https://user-images.githubusercontent.com/37154966/103949151-cf091800-5108-11eb-8eac-fbd82ede93eb.png)

Can lights just NOT be used unless the FOW system is used too?

**Comment by Kruptein** [comment]
Hey, sorry I didn't get a notification for the discussions, so missed this one.

It is currently indeed kinda implied that you use lights in combination with the FOW system.
The effect with the black circle you see is definitely a bug.

I have kept pushing the fix further back because it's somewhat related to something else I would like to fix and that is to provide some proper solution for day scenes with proper token vision.  Right now it's not possible to have a day scene (i.e. fill map with FOW disabled) and have areas that are obscured (e.g. buildings) but work properly with the vision system.

That said, the actual fix for the lights that turn black can be fixed sooner.

**Reply by PalasX** [reply]
No rush, pretty simple solution to just turn off player tokens light sources. Who would walk around in the middle of the day with a lit torch, anyway?! Haha. Just wanted to make sure i wasn't crazy or had somehow turned all the lights transparency into a signed ahort and looped around into negative intensity values :)

**Reply by dthv** [reply]
Actually, it's a native implementation of the *darkness* spell. :P

**Reply by e-kosten** [reply]
For the daylight scene vision, could it be enough to implement in a similar fashion to FOW, but where tokens have "infinite" vision still blocked by walls? I've noticed that with FOW turned on, all tokens have vision in a very small radius around themselves. Not sure how that is implemented, but maybe it would just be a matter of extending the distance of that default.

**Reply by dthv** [reply]
Yes, this would work. But it would work only with locations that are outside-only.But the moment you'd walk into a building without lights and windows, you'd still see everything that is not blocked by walls despite there being no light.

The next step would be to check whether one was inside or outside a closed polygon of walls. Yet, this would break the possibility of having enclosed courts or walled gardens.

What's needed to solve the "daylight/outside location" problem is a method that can differentiate between "inside" and "outside" that is (a) easy to use and (b) light in computation to not slow down/break users' browsers.

**Reply by e-kosten** [reply]
Good point about inside/outside. Would it be desirable for daylight to extend into darkened spaces? For instance, at the opening of a cave, sunlight would allow you to see part way into the cave, but after a point it would be total darkness.

**Reply by Kruptein** [reply]
It would be amazing, question is, is it feasible :p

**Reply by e-kosten** [reply]
Is it possible to implement a line as a light source, rather than a point? If so, potentially we could create such a light source at the border between outside/inside.

**Reply by dthv** [reply]
That should, in principle, be possible, I think.

Yet, the system has to be able to properly distinguish between "inside" and "outside", first. *That's* the hard part for now afaik.

**Reply by e-kosten** [reply]
Can we make this something explicit that is drawn on the FOW layer? Something like a "ceiling" object.  Then, there is no need to make implicit assumptions about anything like whether a closed polygon is a covered building (inside) or an open walled garden (outside).

**Reply by dthv** [reply]
Yes, in principle, we can. When I introduced this idea to kruptein, they said something along the lines of additional layers being heavy in computing resources. On the other hands one would not always use that layer.

Another idea was to use walls that have a distinct "inside" and "outside" side; mathematically, I think this could be done with vectors.

Keep on throwing ideas around. :)

**Reply by PalasX** [reply]
rounding back to the infinite vision distance idea, this could possibly work if lack of vision only obscured the token layer, and the map layer was always 100% visible? The players would always be able to see the entirety of the map, but wouldn't know about the 35 kobold tokens i stuffed into the broom closet of death :)

**Reply by dthv** [reply]
That is, strictly speaking, another problem that is (eventually) to be covered in issue #317 

It also does not solve the problem of combining light and dark areas.

**Reply by e-kosten** [reply]
Would it need to be a whole new layer? I was more just thinking as an object in the existing FOW layer. Combined with the infinite vision, what if in outdoor settings vision rays were infinite in length, but end upon "contact" with one of these "inside" objects? So if you're inside one of those objects, the infinite vision would stop immediately (reverting to vision by auras) and when outside, the vision would stop at either a wall or an inside area?

**Reply by dthv** [reply]
Uh, that's actually nice. "define inside area" as a bool for a shape to let it switch… :thinking:

It would be a bit of a hassle UI-wise, because it would only be applicable to *closed* shapes and we would need to find an intuitive method to have it stack with walls without being obstrusive. We would, also, have the problem of doors, where interaction with other objects would be necessary.

But as far as *I* understand the system, it *might* just work… :thinking:

**Reply by Kruptein** [reply]
I'll take time this weekend to go through this discussion :D

**Reply by dthv** [reply]
(I generally should replace all "we"s with "kruptein" and hope they get not pinged anytime I do. Would be so much more honest. :grin: )

**Reply by Kruptein** [reply]
Some good ideas being thrown around.
> Something like a "ceiling" object.

This is interesting because it's something I've been toying with myself but for other reasons.

> additional layers being heavy in computing resources

I'm starting to warm up more to this idea, given that they indeed don't need to actively be rendered at all times compared to the fow layer.  The nice thing about separate layers is that it has clear intent, there is less risk of confusion compared to say one layer that contains multiple objects but different interactions.  The one downside is that it does cause some decoupling, for example walls drawn on the fow layer right now won't move along with your map if you move it.  But this is a different topic entirely :D

I'll see if I can find time to focus on some brainstorming time with all these ideas!

**Reply by dthv** [reply]
One thing that has to be taken into account with "ceiling" objects is *height*.

With the introduction of *floors*, PA has gone vertical. Yet, if I understand correctly, floors are calculated *horizontally*, i.e. walls block vision on the current floor but are ignored for heigher floors. With "ceiling"-objects, this is problematic, because anything *below* a ceiling would be considered "indoors", while one could actually stand *on* a ceiling (i.e. roof) which would be *outdoors*. Also, one would expect to be able to look down on roofs from heigher buildings.

Thus, if we integrate additional layers, these would need to be coupled to floors, to allow buildings of different heights on the same map and also to allow standing on roofs. This, in turn would mean the introduction of up to *n* additional layers, *n* equalling the number of floors on a location.

Meh.
